File: brdgme-go/roll_through_the_ages/render.go
package roll_through_the_ages

import (
	"bytes"
	"fmt"
	"strconv"
	"strings"

	"github.com/brdgme/brdgme/brdgme-go/render"
)

func (g *Game) RenderName(player int) string {
	return fmt.Sprintf("{{player %d}}", player)
}

func (g *Game) PlayerRender(player int) string {
	buf := bytes.NewBuffer([]byte{})
	// Dice
	diceRow := []render.Cell{}
	numberRow := []render.Cell{}
	for i, d := range g.RolledDice {
		diceString := DiceStrings[d]
		diceRow = append(diceRow, render.Cel(render.Bold(RenderDice(d))))
		numberRow = append(numberRow, render.Cel(fmt.Sprintf(
			`%s%v`,
			strings.Repeat(" ", len(diceString)/2),
			render.Fg(render.Grey, strconv.Itoa(i+1)),
		)))
	}
	for _, d := range g.KeptDice {
		diceRow = append(diceRow, render.Cell{Align: render.Left, Content: RenderDice(d)})
	}
	buf.WriteString("{{b}}Dice{{/b}} ")
	buf.WriteString(render.Fg(render.Grey, "(F: food, W: worker, G: good, C: coin, X: skull)"))
	buf.WriteString("\n")
	t := render.Table([][]render.Cell{diceRow, numberRow}, 0, 2)
	buf.WriteString(t)
	buf.WriteString("\n\n")
	// Remaining turns
	if g.FinalRound {
		buf.WriteString("{{b}}This is the final round{{/b}}\n\n")
	}
	// Turn resources
	switch g.Phase {
	case PhaseBuild, PhaseBuy:
		cells := [][]render.Cell{
			{render.Cel(render.Bold("Turn supplies"))},
			{render.Cel(render.Bold("Workers:")), render.Cel(strconv.Itoa(g.RemainingWorkers))},
			{render.Cel(render.Bold("Coins:")), render.Cel(fmt.Sprintf(
				"%d (%d including goods)",
				g.RemainingCoins,
				g.RemainingCoins+g.Boards[player].GoodsValue(),
			))},
		}
		buf.WriteString(render.Table(cells, 0, 2))
		buf.WriteString("\n\n")
	case PhaseTrade:
		cells := [][]render.Cell{
			{render.Cel(render.Bold("Turn supplies"))},
			{render.Cel(render.Bold("Ships:")), render.Cel(strconv.Itoa(g.RemainingShips))},
		}
		buf.WriteString(render.Table(cells, 0, 2))
		buf.WriteString("\n\n")
	}
	// Cities
	buf.WriteString("{{b}}Cities{{/b}} ")
	buf.WriteString(render.Fg(render.Grey, "(number of dice and food used per turn)"))
	buf.WriteString("\n")
	cityHeaderBuf := bytes.NewBufferString(fmt.Sprintf(
		"{{b}}%d{{/b}}", BaseCitySize))
	last := 0
	for i, n := range CityLevels {
		cityHeaderBuf.WriteString(fmt.Sprintf(
			`%s{{b}}%d{{/b}}`,
			strings.Repeat(" ", (n-last-1)*2+1),
			BaseCitySize+i+1,
		))
		last = n
	}
	cells := [][]render.Cell{{render.Cel("{{b}}Player{{/b}}"), render.Cel(cityHeaderBuf.String())}}
	for p := 0; p < g.PlayerCount(); p++ {
		remaining := MaxCityProgress - g.Boards[p].CityProgress
		row := []render.Cell{
			render.Cel(g.RenderName(p)),
			render.Cel(fmt.Sprintf(
				"%s%s",
				strings.Repeat(fmt.Sprintf(
					`%s `,
					RenderX(p, p == player),
				), g.Boards[p].CityProgress+1),
				strings.Repeat(
					render.Fg(render.Grey, ".")+" ",
					remaining,
				),
			)),
		}
		if remaining > 0 {
			row = append(row, render.Cel(render.Markup(
				fmt.Sprintf("(%d left)", remaining), render.Grey, p == player)))
		}
		cells = append(cells, row)
	}
	t = render.Table(cells, 0, 2)
	buf.WriteString(t)
	buf.WriteString("\n\n")
	// Developments
	header := []render.Cell{render.Cel(render.Bold("Development"))}
	for p := 0; p < g.PlayerCount(); p++ {
		header = append(header, render.Cel(g.RenderName(p)))
	}
	header = append(header, []render.Cell{
		render.Cel(render.Bold("Cost")),
		render.Cel(render.Bold("Pts")),
		render.Cel(render.Bold("Effect")),
	}...)
	cells = [][]render.Cell{header}
	for _, d := range Developments {
		dv := DevelopmentValues[d]
		row := []render.Cell{render.Cel(strings.Title(dv.Name))}
		for p := 0; p < g.PlayerCount(); p++ {
			cell := render.Fg(render.Grey, ".")
			if g.Boards[p].Developments[d] {
				cell = RenderX(p, player == p)
			}
			row = append(row, render.Cel(cell, render.Center))
		}
		row = append(row, []render.Cell{
			render.Cel(fmt.Sprintf(" %d", dv.Cost)),
			render.Cel(fmt.Sprintf(" %d", dv.Points)),
			render.Cel(render.Fg(render.Grey, dv.Effect)),
		}...)
		cells = append(cells, row)
	}
	t = render.Table(cells, 0, 2)
	buf.WriteString(t)
	buf.WriteString("\n\n")
	// Monuments
	header = []render.Cell{render.Cel(render.Bold("Monument"))}
	for p := 0; p < g.PlayerCount(); p++ {
		header = append(header, render.Cel(g.RenderName(p)))
	}
	header = append(header, []render.Cell{
		render.Cel(render.Bold("Size")),
		render.Cel(render.Bold("Pts")),
		render.Cel(render.Bold("Effect")),
	}...)
	cells = [][]render.Cell{header}
	for _, m := range Monuments {
		mv := MonumentValues[m]
		row := []render.Cell{render.Cel(strings.Title(mv.Name))}
		for p := 0; p < g.PlayerCount(); p++ {
			var cell string
			switch {
			case g.Boards[p].Monuments[m] == 0:
				cell = render.Fg(render.Grey, ".")
			case g.Boards[p].Monuments[m] == mv.Size:
				cell = RenderX(p, g.Boards[p].MonumentBuiltFirst[m])
			default:
				cell = render.Fgp(p, strconv.Itoa(g.Boards[p].Monuments[m]))
			}
			row = append(row, render.Cel(cell, render.Center))
		}
		row = append(row, []render.Cell{
			render.Cel(fmt.Sprintf(" %d", mv.Size)),
			render.Cel(fmt.Sprintf("{{b}}%d{{/b}}/%d", mv.Points, mv.SubsequentPoints)),
			render.Cel(render.Fg(render.Grey, mv.Effect)),
		}...)
		cells = append(cells, row)
	}
	t = render.Table(cells, 0, 2)
	buf.WriteString(t)
	buf.WriteString("\n\n")
	// Resources
	header = []render.Cell{render.Cel(render.Bold("Resource"))}
	for p := 0; p < g.PlayerCount(); p++ {
		header = append(header, render.Cel(g.RenderName(p)))
	}
	cells = [][]render.Cell{header}
	for _, good := range GoodsReversed() {
		row := []render.Cell{render.Cel(RenderGoodName(good))}
		for p := 0; p < g.PlayerCount(); p++ {
			num := g.Boards[p].Goods[good]
			cell := render.Fg(render.Grey, ".")
			if num > 0 {
				cell = render.BoldIf(
					render.Fgp(p, fmt.Sprintf("%d (%d)", num, GoodValue(good, num))),
					p == player,
				)
			}
			row = append(row, render.Cel(cell, render.Center))
		}
		cells = append(cells, row)
	}
	row := []render.Cell{render.Cel(render.Bold("total"))}
	for p := 0; p < g.PlayerCount(); p++ {
		cell := render.BoldIf(
			render.Fgp(p, fmt.Sprintf("%d (%d)", g.Boards[p].GoodsNum(), g.Boards[p].GoodsValue())),
			p == player,
		)
		row = append(row, render.Cel(cell, render.Center))
	}
	cells = append(cells, row, []render.Cell{})

	row = []render.Cell{render.Cel(FoodName)}
	for p := 0; p < g.PlayerCount(); p++ {
		cell := render.BoldIf(
			render.Fgp(p, strconv.Itoa(g.Boards[p].Food)),
			p == player,
		)
		row = append(row, render.Cel(cell, render.Center))
	}
	cells = append(cells, row)
	row = []render.Cell{render.Cel(ShipName)}
	for p := 0; p < g.PlayerCount(); p++ {
		cell := render.BoldIf(
			render.Fgp(p, strconv.Itoa(g.Boards[p].Ships)),
			p == player,
		)
		row = append(row, render.Cel(cell, render.Center))
	}
	cells = append(cells, row)
	row = []render.Cell{render.Cel(DisasterName)}
	for p := 0; p < g.PlayerCount(); p++ {
		cell := render.BoldIf(
			render.Fgp(p, strconv.Itoa(g.Boards[p].Disasters)),
			p == player,
		)
		row = append(row, render.Cel(cell, render.Center))
	}
	cells = append(cells, row)
	row = []render.Cell{render.Cel(render.Bold("score"))}
	for p := 0; p < g.PlayerCount(); p++ {
		cell := render.BoldIf(
			render.Fgp(p, strconv.Itoa(g.Boards[p].Score())),
			p == player,
		)
		row = append(row, render.Cel(cell, render.Center))
	}
	cells = append(cells, row)

	t = render.Table(cells, 0, 2)
	buf.WriteString(t)
	return buf.String()
}

func (g *Game) PubRender() string {
	// No hidden information, so the public render is just the render for the
	// active player
	return g.PlayerRender(g.CurrentPlayer)
}

func RenderX(player int, strong bool) string {
	x := "x"
	if strong {
		x = "X"
	}
	return render.BoldIf(render.Fgp(player, x), strong)
}

func RenderDice(dice Die) string {
	diceString := DiceStrings[dice]
	for v, col := range DiceValueColours {
		diceString = strings.Replace(diceString, v, render.Fg(col, v), -1)
	}
	return diceString
}

func RenderGoodName(good Good) string {
	return render.Bold(render.Fg(GoodColours[good], GoodStrings[good]))
}

var FoodName = render.Bold(render.Fg(render.Green, "food"))
var ShipName = render.Bold(render.Fg(render.Blue, "ship"))
var DisasterName = render.Bold(render.Fg(render.Red, "disaster"))

File: rust/game/roll-through-the-ages-2/src/render.rs
//! Port of `brdgme-go/roll_through_the_ages_1/render.go`.
//!
//! No hidden information in this game: Go's `PubRender()` is literally
//! `PlayerRender(CurrentPlayer)` - `PubState`'s render below does exactly
//! that, not a neutral/different layout.
//!
//! Every `render.Table(cells, 0, 2)` call in the Go source uses
//! `colSpacing=2`; all six tables here use `table_with_gap(&rows, 2)` to
//! insert the equivalent spacer cells (see `docs/porting/RENDER_PARITY.md`).

use brdgme_color::NamedColor;
use brdgme_game::game::Renderer;
use brdgme_markup::ast::Cell;
use brdgme_markup::{Align as A, Node as N, Row, table_with_gap};

use crate::development::DEVELOPMENTS;
use crate::dice::Die;
use crate::good::{Good, goods_reversed};
use crate::monument::MONUMENTS;
use crate::player_board::{BASE_CITY_SIZE, CITY_LEVELS, MAX_CITY_PROGRESS};
use crate::{Game, Phase, PlayerState, PubState};

/// Port of `RenderX`.
fn render_x(player: usize, strong: bool) -> N {
    let x = if strong { "X" } else { "x" };
    let coloured = N::Fg(player.into(), vec![N::text(x)]);
    if strong {
        N::Bold(vec![coloured])
    } else {
        coloured
    }
}

/// Port of `render.BoldIf`.
fn bold_if(node: N, when: bool) -> N {
    if when { N::Bold(vec![node]) } else { node }
}

/// Port of `render.Fgp` for arbitrary text.
fn fgp(player: usize, content: impl Into<String>) -> N {
    N::Fg(player.into(), vec![N::text(content.into())])
}

/// Port of `RenderDice`: colours each occurrence of a known letter within
/// the die's face string (`DiceValueColours`), grouping consecutive
/// same-coloured characters into a single run for a smaller node tree
/// (equivalent output to Go's per-letter `strings.Replace`).
fn render_dice(die: Die) -> Vec<N> {
    let s = die.face_string();
    let mut nodes = vec![];
    let mut run = String::new();
    let mut run_colour = None;
    for c in s.chars() {
        let colour = crate::dice::dice_value_colour(c);
        if colour != run_colour {
            if !run.is_empty() {
                nodes.push(match run_colour {
                    Some(col) => N::Fg(col.into(), vec![N::text(run.clone())]),
                    None => N::text(run.clone()),
                });
            }
            run.clear();
            run_colour = colour;
        }
        run.push(c);
    }
    if !run.is_empty() {
        nodes.push(match run_colour {
            Some(col) => N::Fg(col.into(), vec![N::text(run)]),
            None => N::text(run),
        });
    }
    nodes
}

/// Port of `RenderGoodName`.
fn render_good_name(good: Good) -> N {
    N::Bold(vec![N::Fg(
        good.colour().into(),
        vec![N::text(good.name())],
    )])
}

/// Port of `strings.Title` as applied to lowercase, space-separated names
/// (development names are single lowercase words; monument names are
/// already Title Case, so this is idempotent for them).
fn title_case(s: &str) -> String {
    s.split(' ')
        .map(|word| {
            let mut chars = word.chars();
            match chars.next() {
                Some(first) => first.to_uppercase().collect::<String>() + chars.as_str(),
                None => String::new(),
            }
        })
        .collect::<Vec<_>>()
        .join(" ")
}

fn cell(content: Vec<N>) -> Cell {
    (A::Left, content)
}

fn cell_align(content: Vec<N>, align: A) -> Cell {
    (align, content)
}

/// Port of `Game.PlayerRender`.
fn render(game: &Game, player: usize) -> Vec<N> {
    let mut out: Vec<N> = vec![];

    // Dice
    let mut dice_row: Row = vec![];
    let mut number_row: Row = vec![];
    for (i, &d) in game.rolled_dice.iter().enumerate() {
        dice_row.push(cell(vec![N::Bold(render_dice(d))]));
        let dice_string = d.face_string();
        number_row.push(cell(vec![
            N::text(" ".repeat(dice_string.len() / 2)),
            N::Fg(NamedColor::Grey.into(), vec![N::text((i + 1).to_string())]),
        ]));
    }
    for &d in game.kept_dice.iter() {
        dice_row.push(cell(render_dice(d)));
    }
    out.push(N::Bold(vec![N::text("Dice")]));
    out.push(N::text(" "));
    out.push(N::Fg(
        NamedColor::Grey.into(),
        vec![N::text("(F: food, W: worker, G: good, C: coin, X: skull)")],
    ));
    out.push(N::text("\n"));
    out.push(table_with_gap(&[dice_row, number_row], 2));
    out.push(N::text("\n\n"));

    // Remaining turns
    if game.final_round {
        out.push(N::Bold(vec![N::text("This is the final round")]));
        out.push(N::text("\n\n"));
    }

    // Turn resources
    match game.phase {
        Phase::Build | Phase::Buy => {
            let cells: Vec<Row> = vec![
                vec![cell(vec![N::Bold(vec![N::text("Turn supplies")])])],
                vec![
                    cell(vec![N::Bold(vec![N::text("Workers:")])]),
                    cell(vec![N::text(game.remaining_workers.to_string())]),
                ],
                vec![
                    cell(vec![N::Bold(vec![N::text("Coins:")])]),
                    cell(vec![N::text(format!(
                        "{} ({} including goods)",
                        game.remaining_coins,
                        game.remaining_coins + game.boards[game.current_player].goods_value(),
                    ))]),
                ],
            ];
            out.push(table_with_gap(&cells, 2));
            out.push(N::text("\n\n"));
        }
        Phase::Trade => {
            let cells: Vec<Row> = vec![
                vec![cell(vec![N::Bold(vec![N::text("Turn supplies")])])],
                vec![
                    cell(vec![N::Bold(vec![N::text("Ships:")])]),
                    cell(vec![N::text(game.remaining_ships.to_string())]),
                ],
            ];
            out.push(table_with_gap(&cells, 2));
            out.push(N::text("\n\n"));
        }
        _ => {}
    }

    // Cities
    out.push(N::Bold(vec![N::text("Cities")]));
    out.push(N::text(" "));
    out.push(N::Fg(
        NamedColor::Grey.into(),
        vec![N::text("(number of dice and food used per turn)")],
    ));
    out.push(N::text("\n"));
    let mut city_header: Vec<N> = vec![N::Bold(vec![N::text(BASE_CITY_SIZE.to_string())])];
    let mut last = 0;
    for (i, &n) in CITY_LEVELS.iter().enumerate() {
        city_header.push(N::text(" ".repeat(((n - last - 1) * 2 + 1) as usize)));
        city_header.push(N::Bold(vec![N::text(
            (BASE_CITY_SIZE + i as i32 + 1).to_string(),
        )]));
        last = n;
    }
    let mut cities_rows: Vec<Row> = vec![vec![
        cell(vec![N::Bold(vec![N::text("Player")])]),
        cell(city_header),
    ]];
    for p in 0..game.players {
        let remaining = MAX_CITY_PROGRESS - game.boards[p].city_progress;
        let marker = render_x(p, p == player);
        let mut progress_nodes: Vec<N> = vec![];
        for _ in 0..(game.boards[p].city_progress + 1) {
            progress_nodes.push(marker.clone());
            progress_nodes.push(N::text(" "));
        }
        for _ in 0..remaining {
            progress_nodes.push(N::Fg(NamedColor::Grey.into(), vec![N::text(".")]));
            progress_nodes.push(N::text(" "));
        }
        let mut row: Row = vec![cell(vec![N::Player(p)]), cell(progress_nodes)];
        if remaining > 0 {
            let left = bold_if(
                N::Fg(
                    NamedColor::Grey.into(),
                    vec![N::text(format!("({} left)", remaining))],
                ),
                p == player,
            );
            row.push(cell(vec![left]));
        }
        cities_rows.push(row);
    }
    out.push(table_with_gap(&cities_rows, 2));
    out.push(N::text("\n\n"));

    // Developments
    let mut dev_header: Row = vec![cell(vec![N::Bold(vec![N::text("Development")])])];
    for p in 0..game.players {
        dev_header.push(cell(vec![N::Player(p)]));
    }
    dev_header.push(cell(vec![N::Bold(vec![N::text("Cost")])]));
    dev_header.push(cell(vec![N::Bold(vec![N::text("Pts")])]));
    dev_header.push(cell(vec![N::Bold(vec![N::text("Effect")])]));
    let mut dev_rows: Vec<Row> = vec![dev_header];
    for &d in DEVELOPMENTS.iter() {
        let dv = d.value();
        let mut row: Row = vec![cell(vec![N::text(title_case(dv.name))])];
        for p in 0..game.players {
            let node = if game.boards[p].developments.contains(&d) {
                render_x(p, player == p)
            } else {
                N::Fg(NamedColor::Grey.into(), vec![N::text(".")])
            };
            row.push(cell_align(vec![node], A::Center));
        }
        row.push(cell(vec![N::text(format!(" {}", dv.cost))]));
        row.push(cell(vec![N::text(format!(" {}", dv.points))]));
        row.push(cell(vec![N::Fg(
            NamedColor::Grey.into(),
            vec![N::text(dv.effect)],
        )]));
        dev_rows.push(row);
    }
    out.push(table_with_gap(&dev_rows, 2));
    out.push(N::text("\n\n"));

    // Monuments
    let mut mon_header: Row = vec![cell(vec![N::Bold(vec![N::text("Monument")])])];
    for p in 0..game.players {
        mon_header.push(cell(vec![N::Player(p)]));
    }
    mon_header.push(cell(vec![N::Bold(vec![N::text("Size")])]));
    mon_header.push(cell(vec![N::Bold(vec![N::text("Pts")])]));
    mon_header.push(cell(vec![N::Bold(vec![N::text("Effect")])]));
    let mut mon_rows: Vec<Row> = vec![mon_header];
    for &m in MONUMENTS.iter() {
        let mv = m.value();
        let mut row: Row = vec![cell(vec![N::text(title_case(mv.name))])];
        for p in 0..game.players {
            let progress = game.boards[p].monuments.get(&m).copied().unwrap_or(0);
            let node = if progress == 0 {
                N::Fg(NamedColor::Grey.into(), vec![N::text(".")])
            } else if progress == mv.size {
                render_x(p, game.boards[p].monument_built_first.contains(&m))
            } else {
                fgp(p, progress.to_string())
            };
            row.push(cell_align(vec![node], A::Center));
        }
        row.push(cell(vec![N::text(format!(" {}", mv.size))]));
        row.push(cell(vec![
            N::Bold(vec![N::text(mv.points.to_string())]),
            N::text(format!("/{}", mv.subsequent_points)),
        ]));
        row.push(cell(vec![N::Fg(
            NamedColor::Grey.into(),
            vec![N::text(mv.effect)],
        )]));
        mon_rows.push(row);
    }
    out.push(table_with_gap(&mon_rows, 2));
    out.push(N::text("\n\n"));

    // Resources
    let mut res_header: Row = vec![cell(vec![N::Bold(vec![N::text("Resource")])])];
    for p in 0..game.players {
        res_header.push(cell(vec![N::Player(p)]));
    }
    let mut res_rows: Vec<Row> = vec![res_header];
    for good in goods_reversed().iter() {
        let mut row: Row = vec![cell(vec![render_good_name(*good)])];
        for p in 0..game.players {
            let num = game.boards[p].goods.get(good).copied().unwrap_or(0);
            let node = if num > 0 {
                bold_if(
                    fgp(
                        p,
                        format!("{} ({})", num, crate::good::good_value(*good, num)),
                    ),
                    p == player,
                )
            } else {
                N::Fg(NamedColor::Grey.into(), vec![N::text(".")])
            };
            row.push(cell_align(vec![node], A::Center));
        }
        res_rows.push(row);
    }
    let mut total_row: Row = vec![cell(vec![N::Bold(vec![N::text("total")])])];
    for p in 0..game.players {
        let node = bold_if(
            fgp(
                p,
                format!(
                    "{} ({})",
                    game.boards[p].goods_num(),
                    game.boards[p].goods_value()
                ),
            ),
            p == player,
        );
        total_row.push(cell_align(vec![node], A::Center));
    }
    res_rows.push(total_row);
    res_rows.push(vec![]);

    let mut food_row: Row = vec![cell(vec![N::Bold(vec![N::Fg(
        NamedColor::Green.into(),
        vec![N::text("food")],
    )])])];
    for p in 0..game.players {
        let node = bold_if(fgp(p, game.boards[p].food.to_string()), p == player);
        food_row.push(cell_align(vec![node], A::Center));
    }
    res_rows.push(food_row);

    let mut ship_row: Row = vec![cell(vec![N::Bold(vec![N::Fg(
        NamedColor::Blue.into(),
        vec![N::text("ship")],
    )])])];
    for p in 0..game.players {
        let node = bold_if(fgp(p, game.boards[p].ships.to_string()), p == player);
        ship_row.push(cell_align(vec![node], A::Center));
    }
    res_rows.push(ship_row);

    let mut disaster_row: Row = vec![cell(vec![N::Bold(vec![N::Fg(
        NamedColor::Red.into(),
        vec![N::text("disaster")],
    )])])];
    for p in 0..game.players {
        let node = bold_if(fgp(p, game.boards[p].disasters.to_string()), p == player);
        disaster_row.push(cell_align(vec![node], A::Center));
    }
    res_rows.push(disaster_row);

    let mut score_row: Row = vec![cell(vec![N::Bold(vec![N::text("score")])])];
    for p in 0..game.players {
        let node = bold_if(fgp(p, game.boards[p].score().to_string()), p == player);
        score_row.push(cell_align(vec![node], A::Center));
    }
    res_rows.push(score_row);

    out.push(table_with_gap(&res_rows, 2));

    out
}

impl Renderer for PubState {
    fn render(&self) -> Vec<N> {
        render(&self.game, self.game.current_player)
    }
}

impl Renderer for PlayerState {
    fn render(&self) -> Vec<N> {
        render(&self.game, self.player)
    }
}
